Support protobuf 3.22 or upper

owent opened the pull request:
Fixes #2095 

## Changes

Protobuf depends on abseil-cpp 20230125 from v22/v4.22.

It removed `stringpiece.h`, `google::protobuf::Base64Escape` and use `absl::Base64Escape` for alternative.

~~We have encountered multiple problems related to the usage of CMake packages in the past (#705 #1359). As a solution, I have added a patch script `cmake/patch-imported-config.cmake` that allows users to use prebuilt packages with different CONFIG settings when building otel-cpp.~~

BTW: gRPC v1.55.0 also depend the new protobuf and abseil-cpp, so I also test the new gRPC version in the new CI job.

+ Support protobuf 3.22 or upper
+ Add a CI job (`cmake_modern_protobuf_grpc_test`) to test building with new protobuf and gRPC.
+ ~~Patch imported target in cmake CONFIG packages to enable users to use prebuilt packages with different CONFIG settings.~~
+ Remove `--experimental_allow_proto3_optional` when running protoc when we use protobuf 3.16 or upper. (It's not experimental any more)

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

esigo commented:
@owent could you please add more detailed information in the PR description. thanks :)

owent commented:
> @owent could you please add more detailed information in the PR description. thanks :)

Done.

lalitb commented on ci/setup_grpc.sh:
```diff
@@ -7,24 +7,41 @@ set -e
 export DEBIAN_FRONTEND=noninteractive
 old_grpc_version='v1.33.2'
 new_grpc_version='v1.49.2'
+moderd_grpc_version='v1.55.0'
```
nit - did we mean `modern` ?

owent replied:
Thanks and fixed.

lalitb commented on .github/workflows/ci.yml:
```diff
@@ -387,6 +387,31 @@ jobs:
+      env:
+        PROTOBUF_VERSION: '23.2'
+        ABSEIL_CPP_VERSION: '20230125.3'
+        CXX_STANDARD: '17'
```
Why do we need C++17 standard for build ?

owent replied:
The new versions of abseil-cpp, protobuf and gRPC requires C++14, and this version of abseil-cpp seems has some compatibility problem(See https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/5099430537/jobs/9167125807 for details). So I keep the same setting with gRPC here.

lalitb commented on .github/workflows/ci.yml:
```diff
@@ -387,6 +387,31 @@ jobs:
+        sudo -E ./ci/install_protobuf.sh
+    - name: run otlp exporter tests
+      env:
+        WITH_ABSEIL: 'ON'
```
Should we also have a test without WITH_ABSEIL to ensure that ABI compatible build works fine ?

owent replied:
The latest versions of protobuf and gRPC require the new abseil-cpp, so we must enable it when using these versions.

lalitb replied:
But `WITH_ABSEIL` cmake option will enable `HAVE_ABSEIL` macro, and so enforce to use `abseil::variant` at API surface. Do you mean with latest versions of gRPC and protobuf, we can no longer use `nostd::variant` at API surface ?

lalitb replied:
Thanks for confirming. This looks concerning to me.

owent replied:
Yes, it'a problem.Do you have any suggestion?

lalitb replied:
Not for now. I would like to test the build with newer protobuf to fully understand the issue with our internal nostd::variant, and if we can do any fix in that class. Will try to test during this week. If not, good to discuss this in the next community meeting before deciding.
Thanks once again for providing all the details.

owent replied:
Can we raise another issue to track this ABI compliance problem later? This problem existed before with abseil.

lalitb replied:
Can we wait for Monday SIG meeting, also would like to test this once over weekend to have better understanding. Don't want to hold this PR any way as we need to support latest protobuf.

owent replied:
Thanks.

ThomsonTan commented on CMakeLists.txt:
```diff
@@ -18,6 +18,9 @@ project(opentelemetry-cpp)
 mark_as_advanced(CMAKE_TOOLCHAIN_FILE)
 
+# Prefer cmake CONFIG to auto resolve dependencies.
+set(CMAKE_FIND_PACKAGE_PREFER_CONFIG TRUE)
```
This flag seems new in CMake 3.15, but we support 3.1. Is this required?

marcalff commented on ci/install_protobuf.sh:
```diff
@@ -31,12 +31,41 @@ set -e
+if [[ ${PROTOBUF_VERSION/.*/} -ge 22 ]]; then
+  export CPP_PROTOBUF_VERSION="${PROTOBUF_VERSION}"
+  CPP_PROTOBUF_PACKAGE_NAME="protobuf-${CPP_PROTOBUF_VERSION}"
+  CPP_PROTOBUF_BUILD_OPTIONS=(${CPP_PROTOBUF_BUILD_OPTIONS[@]} "-Dprotobuf_ABSL_PROVIDER=package")
```
Please move this line to the previous test on PROTOBUF_VERSION,
to keep build options together with -Dprotobuf_MODULE_COMPATIBLE=ON

marcalff commented on ci/install_protobuf.sh:
```diff
@@ -31,12 +31,41 @@ set -e
+    CPP_PROTOBUF_BUILD_OPTIONS=(${CPP_PROTOBUF_BUILD_OPTIONS[@]} "-DCMAKE_CXX_STANDARD=${CXX_STANDARD}")
+fi
+
+if [[ ${PROTOBUF_VERSION/.*/} -ge 22 ]]; then
+  CPP_PROTOBUF_BUILD_OPTIONS=(${CPP_PROTOBUF_BUILD_OPTIONS[@]} "-Dprotobuf_MODULE_COMPATIBLE=ON")
+fi
```
Please add a comment to explain _why_ we need this.

Something like:
- protobuf version 22.xxx does ???
- protobuf version 3.yyy does ??? instead

owent replied:
`protobuf_MODULE_COMPATIBLE=ON` is not required now, I remove it now.

marcalff commented on ci/install_protobuf.sh:
```diff
@@ -31,12 +31,41 @@ set -e
+  CPP_PROTOBUF_BUILD_OPTIONS=(${CPP_PROTOBUF_BUILD_OPTIONS[@]} "-Dprotobuf_MODULE_COMPATIBLE=ON")
+fi
+
+if [[ ${PROTOBUF_VERSION/.*/} -ge 22 ]]; then
+  export CPP_PROTOBUF_VERSION="${PROTOBUF_VERSION}"
+  CPP_PROTOBUF_PACKAGE_NAME="protobuf-${CPP_PROTOBUF_VERSION}"
```
Please add a comment to explain _why_ we need this.

Something like:
- protobuf 22.xxx ships packages named ???
- protobuf 3.yyy ships packages named ???

Also, please clarify if we should pick:

https://github.com/protocolbuffers/protobuf/releases/download/v22.3/protobuf-22.3.tar.gz

or

https://github.com/protocolbuffers/protobuf/archive/refs/tags/v22.3.tar.gz

If there is a difference, detail why.

If this is the same, can the following URL pattern work for both 22.xxx and 3.yyy ?

https://github.com/protocolbuffers/protobuf/archive/refs/tags/v22.3.tar.gz

I admit I am totally confused here by the protobuf version names, hence comments explaining what probuf is doing will help.

owent replied:
It seems 3.XX.YY/4.XX.YY is alias for XX.YY. And there will be no 3.XX.YY after 22.
And the released source packages are moved into the tag of XX.YY and not 4.XX.YY after 22.
Also, there is only protobuf-22.3.tar.gz, and there is no protobuf-cpp-22.3.tar.gz any more.

marcalff commented on ci/setup_grpc.sh:
```diff
@@ -7,24 +7,41 @@ set -e
 usage() { echo "Usage: $0 [-v <gcc-version>] [-i <install_dir>"] 1>&2; exit 1;}
 
-while getopts ":v:i:r:s:TH" o; do
+while getopts ":v:i:mp:r:s:TH" o; do
```
Please update the usage() helper to show all options.

owent replied:
Thanks, and all options are added into usage message.

marcalff commented on ci/setup_grpc.sh:
```diff
@@ -59,22 +77,26 @@ git clone --depth=1 -b ${install_grpc_version} https://github.com/grpc/grpc
-)
-if [ ! -z "$build_shared_libs" ]; then
-    ABSEIL_CPP_BUILD_OPTIONS=(${ABSEIL_CPP_BUILD_OPTIONS[@]} "-DBUILD_SHARED_LIBS=$build_shared_libs")
+if [[ $build_internal_abseil_cpp -ne 0 ]]; then
+    mkdir -p "third_party/abseil-cpp/build" && pushd "third_party/abseil-cpp/build"
+    set -x
+
+    ABSEIL_CPP_BUILD_OPTIONS=(
+        -DCMAKE_BUILD_TYPE=Release
+        -DCMAKE_POSITION_INDEPENDENT_CODE=TRUE
+        -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR
+    )
+    if [ ! -z "$build_shared_libs" ]; then
+        ABSEIL_CPP_BUILD_OPTIONS=(${ABSEIL_CPP_BUILD_OPTIONS[@]} "-DBUILD_SHARED_LIBS=$build_shared_libs")
+    fi
+    cmake  ${ABSEIL_CPP_BUILD_OPTIONS[@]} ..
+    cmake --build . -j${nproc} --target install && popd
```
Could all the logic related to abseil be moved to install_abseil.sh instead ?

owent replied:
I'm not sure, but some CI jobs use `setup_grpc.sh` but without `install_abseil.sh`. This installation existed before, I only add a option to allow gRPC find abseil-cpp from a existed installation path.
Maybe we can optimize abseil-cpp installation in another PR?

marcalff reviewed (commented):
Thanks for all the changes.

This fix is extensive, touching a lot of parts in makefiles, and depends on how third party packages are organized, and their own dependencies ...

Obviously, you have a very good understanding of the constraints and technical details.

A reader can see what the CI and install scripts are doing, and this alone requires to reverse engineer the code already to understand what is done.

The critical part which is missing is to know why something is done.

Something like:

protobuf 22 does so-and-so with cmake packages, so as a result we need to use this-and-that to consume it

will go a very long way to grasp what is going on here, and will allow others to actually maintain this.

Please share this knowledge in comments.

lalitb commented:
Have tested the changes, and see relevant. There are 2 points I would like to discuss with this PR
- The internal nostd components are no longer usable for latest version of gRPC and onwards.
- We can't use C++11/14 to build the SDK. The minimum supported version would be  C++17.

marcalff commented:
Looking again at abseil, I don't think we need C++17 with abseil.

Per the abseil doc, C++14 is supported.

Per abseil code, ABSL_USES_STD_STRING_VIEW controls whether std::string_view is used when available in C++17, or whether a replacement is used for C++11/C++14.

If abseil was really broken for std::string_view in C++14, there would be several bug reports and complaints about it, but there are none.

What is more likely here, is that somehow, the CI build does not pass the C++14 standard flags all the way to the compiler, when building code, so that code ends up being compiled assuming C++17, therefore assuming std::string_view is available, causing breaks.

To clarify this, the CI scripts should be fixed to:
* always define a `CXX_STANDARD` from `ci.yaml`
* always use `sudo -E` instead of `sudo`, to propagate the compiler choice (CXX) and C++ mode used
* always use the `CXX_STANDARD` passed in to build a proper `CMAKE_CXX_STANDARD` option, when invoking cmake from various setup_xxx.sh helpers.

Once all this is done, it is likely the code will also build in C++14 mode, with the same `CXX` compiler and the same `C++14` mode used across all build scripts.

owent commented:
> Looking again at abseil, I don't think we need C++17 with abseil.
> 
> Per the abseil doc, C++14 is supported.
> 
> Per abseil code, ABSL_USES_STD_STRING_VIEW controls whether std::string_view is used when available in C++17, or whether a replacement is used for C++11/C++14.
> 
> If abseil was really broken for std::string_view in C++14, there would be several bug reports and complaints about it, but there are none.
> 
> What is more likely here, is that somehow, the CI build does not pass the C++14 standard flags all the way to the compiler, when building code, so that code ends up being compiled assuming C++17, therefore assuming std::string_view is available, causing breaks.
> 
> To clarify this, the CI scripts should be fixed to:
> 
> * always define a `CXX_STANDARD` from `ci.yaml`
> * always use `sudo -E` instead of `sudo`, to propagate the compiler choice (CXX) and C++ mode used
> * always use the `CXX_STANDARD` passed in to build a proper `CMAKE_CXX_STANDARD` option, when invoking cmake from various setup_xxx.sh helpers.
> 
> Once all this is done, it is likely the code will also build in C++14 mode, with the same `CXX` compiler and the same `C++14` mode used across all build scripts.

`CXX_STANDARD` is changed to 14 now.

marcalff commented:
Adding CVE tag: this fix is now required to upgrade to gRPC >= 1.53.0 to fix:
* https://github.com/open-telemetry/opentelemetry-cpp/issues/2231

marcalff commented:
@owent 

Please merge main into this PR to resolve conflicts, so we get up-to-date code.

Thanks.

owent commented:
> @owent
> 
> Please merge main into this PR to resolve conflicts, so we get up-to-date code.
> 
> Thanks.

Done

esigo commented:
Discussed in the weekly 17.07.2023. @lalitb and @marcalff will review. the issue with not being able to build with internal abseil will be handled in a separate issue.

marcalff reviewed (approved):
@owent 

Sorry the review took so long.

Overall comments below

1)

In `otlp_http_client.cc`, there is now an optional dependency on abseil,
for base 64 encoding.

Ok.

2)

In the OTLP exporter CMakeLists, there is an optional dependency on abseil.

Ok.

3)

In the OTLP exporter BUILD, for bazel, the dependency on abseil is required.

This is not optional in bazel, even when the C++ code might not even use abseil.

It would be better to have an optional dependency in the bazel build,
but I am not even sure if bazel supports this to begin with,
and know even less how to implement this ... so no solution to offer here.

Given that the C++ code in otlp_http_client.cc still include bazel optionally,
I guess having a mandatory-but-unused dependency on `absl/strings` from the BUILD file will be ok.

4)

About the changes in `cmake/*.cmake`,
I don't understand CMake well enough to make valuable comments in this area.

5)

Changes in the CI build itself looks ok to me, and beside, these only affects the CI build,
not the end user when using a release, so I think there is little technical risk here.

Having explicit `CXX_STANDARD` lines in `ci.yml` is much better, as the CI code is now clearer, and easier to maintain.

6)

As discussed previously,

when the build uses WITH_ABSEIL because of a dependency,
a side effect is also that the ABI is changed to also use WITH_ABSEIL,
at least in my understanding (to investigate further).

We may need to decouple:

* usage of WITH_ABSEIL in the implementation (protobuf)
* usage of WITH_ABSEIL in the ABI

This will be investigated in a separate issue and PR.

---

LGTM overall.

Thanks for the fix, approved.

owent commented on CHANGELOG.md:
```diff
@@ -170,7 +173,7 @@ Breaking changes:
 * [CI] Upgrade Google Benchmark version from 1.6.0 to 1.7.1
   [#2116](https://github.com/open-telemetry/opentelemetry-cpp/pull/2116)
-* [CI] Upgrade Nlohmann JSON library version from 3.10.5 to 3.11.2
+* [CI] Upgrade Nlohmann JSON library version from 3.10.5 to 3.11.
```
Thanks, this is a mistake.I have restored this change.

owent commented:
> @owent
> 
> Sorry the review took so long.
> 
> Overall comments below
> 
> 
> In `otlp_http_client.cc`, there is now an optional dependency on abseil, for base 64 encoding.
> 
> Ok.
> 
> 
> In the OTLP exporter CMakeLists, there is an optional dependency on abseil.
> 
> Ok.
> 
> 
> In the OTLP exporter BUILD, for bazel, the dependency on abseil is required.
> 
> This is not optional in bazel, even when the C++ code might not even use abseil.
> 
> It would be better to have an optional dependency in the bazel build, but I am not even sure if bazel supports this to begin with, and know even less how to implement this ... so no solution to offer here.
> 
> Given that the C++ code in otlp_http_client.cc still include bazel optionally, I guess having a mandatory-but-unused dependency on `absl/strings` from the BUILD file will be ok.
> 
> 
> About the changes in `cmake/*.cmake`, I don't understand CMake well enough to make valuable comments in this area.
> 
> 
> Changes in the CI build itself looks ok to me, and beside, these only affects the CI build, not the end user when using a release, so I think there is little technical risk here.
> 
> Having explicit `CXX_STANDARD` lines in `ci.yml` is much better, as the CI code is now clearer, and easier to maintain.
> 
> 
> As discussed previously,
> 
> when the build uses WITH_ABSEIL because of a dependency, a side effect is also that the ABI is changed to also use WITH_ABSEIL, at least in my understanding (to investigate further).
> 
> We may need to decouple:
> 
> * usage of WITH_ABSEIL in the implementation (protobuf)
> * usage of WITH_ABSEIL in the ABI
> 
> This will be investigated in a separate issue and PR.
> 
> LGTM overall.
> 
> Thanks for the fix, approved.

3. I not a professional of bazel, but can we use `select()` to switch the implementations of abseil-cpp?
4. The changes in `cmake/*.cmake` optimize some code reuse for protocols and remove the useless options of protoc and the more important change is fix the linking problems when the configure types of third party packages are different from building otel-cpp.
5. I'm not sure if it's easy to decouple the usage of WITH_ABSEIL in the implementation of OTLP exporters. But we can hide the symbols which use protobuf and do not export them to users.

ThomsonTan commented on ci/install_abseil.sh:
```diff
@@ -3,19 +3,27 @@
 # SPDX-License-Identifier: Apache-2.0
 
-set -e
+set -ex
```
Do we need the `-x` option seems which seems for debugging?

ThomsonTan commented on CMakeLists.txt:
```diff
@@ -337,9 +342,21 @@ if(WITH_PROMETHEUS)
+  if(Protobuf_VERSION AND Protobuf_VERSION VERSION_GREATER_EQUAL "3.22.0")
+    if(NOT WITH_ABSEIL)
+      message(
+        FATAL_ERROR
```
Seems with new protobuf, the main SDK will no longer buildable with `WITH_ABSEIL=0` so this will be a required dependency. Can we separate the abseil dependency to protobuf related library only?

owent replied:
protobuf has a internal reference to abseil and will use it when building without `-Dprotobuf_ABSL_PROVIDER=package`, just like gRPC.I think it's better to keep the same version for protobuf, gRPC and otel-cpp.

lalitb commented:
Sorry for the late reply to this. The changes for protobuf upgrade look good. We do have an issue with ABSEIL dependency for the API, but let's track it separately, as it is essential to upgrade with vulnerability in the older protobuf version.
The changes in cmake/*cmake are hard to understand for me too, and seems they are also not directly related to the protobuf upgrade. 

@owent Do you think we can separate out the cmake/*cmake related changes from this PR, and approve this solely for the protobuf upgrade? We can then target to merge it within this week if you can do these changes soon. Or else, if any of the approvers/maintainers are comfortable with the cmake changes here, I am fine with that too :) Sorry once again for lots of delays and asking for more changes.

owent commented:
@lalitb I create #2251 to track the problems fixed by `cmake/*` , and I will push the codes later after this PR is merged.

lalitb commented:
> @lalitb I create #2251 to track the problems fixed by `cmake/*` , and I will push the codes later after this PR is merged.

Thanks @owent. Will be reviewing it tomorrow. Had a glance, and the changes look good to me in general.

lalitb reviewed (approved):
The changes look good. Thanks for working on this. Couple of concerns in general to investigate separately:
 - Need to have a Github to investigate building "otel-cpp + OTLP exporter" with the latest gRPC/protobuf without the WITH_ABSEIL option. The aim is to enable the API interface of otel-cpp to function with the internal `nostd` components independently, without being reliant on the abseil dependency of protobuf.
 - upgrade the gRPC version for the bazel build.

owent commented on .github/workflows/ci.yml:
```diff
@@ -378,6 +385,31 @@ jobs:
+        CXX_STANDARD: '14'
+      run: |
+        sudo ./ci/setup_cmake.sh
+        sudo ./ci/setup_ci_environment.sh
```
These two script do not use these environment variables. Do we still add `-E`?

> Need to have a Github to investigate building "otel-cpp + OTLP exporter" with the latest gRPC/protobuf without the WITH_ABSEIL option. The aim is to enable the API interface of otel-cpp to function with the internal nostd components independently, without being reliant on the abseil dependency of protobuf.

Done.

> upgrade the gRPC version for the bazel build.

This PR do not upgrade the default version of gRPC and protobuf, just add supporting for the new versions. Do you think we can keep the default versions of gRPC and protobuf unchanged to check the compatibility with the old versions?

lalitb replied:
> This PR do not upgrade the default version of gRPC and protobuf, just add supporting for the new versions. Do you think we can keep the default versions of gRPC and protobuf unchanged to check the compatibility with the old versions?

Yeah good point. Should be fine. thanks.

lalitb replied:
> These two script do not use these environment variables. Do we still add -E?

 Just thought of following a practice if in future we change these scripts to use env variable. But no strong opinion, we can change it later when situation arises.

ThomsonTan commented:
There seems some syntax error for modern protobuf without abseil. Is this expected, or modern protobuf does require abseil?

https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/5753570940/job/15599220018?pr=2163

lalitb commented on exporters/otlp/src/otlp_http_client.cc:
```diff
@@ -65,6 +50,146 @@ namespace otlp
+  Base64EscapeInternal(*dest, reinterpret_cast<const unsigned char *>(src.data()), src.size(),
+                       kBase64CharsBasic, '=');
+}
+
```
Just curious, why to implement this and not use from abseil/google. Maintaining this code would be issue in future ?

lalitb commented on exporters/otlp/src/otlp_http_client.cc:
```diff
@@ -65,6 +50,146 @@ namespace otlp
+  Base64EscapeInternal(*dest, reinterpret_cast<const unsigned char *>(src.data()), src.size(),
+                       kBase64CharsBasic, '=');
+}
+
```
Ok I think we need these methods for building without WITH_ABSEIL and latest protobuf. I hope we have enough confidence with these methods, and tested well. And these are good to use even when `absl::Base64Escape` or `google::protobuf::Base64Escape` are available ? Can we add some unit-tests to test possible scenarios. Thanks.

owent replied:
Sure. These codes are from out internal repo. I will also copy the unit tests here too.

lalitb commented on exporters/otlp/src/otlp_http_client.cc:
```diff
@@ -65,6 +50,146 @@ namespace otlp
+
+  n = (slen + 2) / 3;
+
+  if (n > (BASE64_SIZE_T_MAX - 1) / 4)
```
where is BASE64_SIZE_T_MAX  defined ?

lalitb commented on exporters/otlp/src/otlp_http_client.cc:
```diff
@@ -65,6 +50,146 @@ namespace otlp
+// Encodes a `src` string into a base64-encoded 'dest' string with padding
+// characters. This function conforms with RFC 4648 section 4 (base64) and RFC
+// 2045. See also CalculateBase64EscapedLen().
+static void Base64Escape(nostd::string_view src, std::string *dest)
```
should `dest` by reference ?

owent replied:
I think it's better to keep the same parameters as it in abseil.

lalitb commented on exporters/otlp/src/otlp_http_client.cc:
```diff
@@ -63,6 +51,146 @@ namespace otlp
 namespace
 {
 
+using Base64EscapeChars                              = const unsigned char[64];
```
should these methods be moved to https://github.com/open-telemetry/opentelemetry-cpp/blob/main/api/include/opentelemetry/common/string_util.h ?

owent replied:
There are a lot of codes here, do you think it's better to move it into sdk?

lalitb replied:
Yeah good point, not required. We can move it later if need to share with other components. Thanks.

owent replied:
base64 algorithm is moved into sdk now.

lalitb replied:
Really sorry if my comment above was not clear. I did realize that this will bloat the sdk, and we should move it only when we need to share it with other components.

lalitb commented on exporters/otlp/src/otlp_http_client.cc:
```diff
@@ -63,6 +51,146 @@ namespace otlp
+  {
+    return;
+  }
+
```
early exit if `src` is empty ?

owent commented:
> * Need to have a Github task to investigate building "otel-cpp + OTLP exporter" with the latest gRPC/protobuf without the WITH_ABSEIL option. The aim is to enable the API interface of otel-cpp to function with the internal `nostd` components independently, without being reliant on the abseil dependency of protobuf.
> * upgrade the gRPC version for the bazel build.

I just tested this CI job and find protobuf can not work when both external and internal abseil-cpp exists.Because they share the same header paths. Maybe we can move headers os internal abseil-cpp into another path so we can keep ABI compatibility and use the internal and external abseil at the same time?

@lalitb This PR is ready to be reviewed again.

owent commented:
> There seems some syntax error for modern protobuf without abseil. Is this expected, or modern protobuf does require abseil?
> 
> https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/5753570940/job/15599220018?pr=2163

I just tested this CI job and find protobuf can not work when both external and internal abseil-cpp exists.Because they share the same header paths. Maybe we can move headers os internal abseil-cpp into another path so we can keep ABI compatibility and use the internal and external abseil at the same time?

@lalitb This PR is ready to be reviewed again and thanks,

lalitb commented:
> Maybe we can move headers os internal abseil-cpp into another path, do not inline namespace otel_v1 and replace all usage absl:: in otel-cpp by a macro (OTABSL_NAMESPACE for example) and then we can define OTABSL_NAMESPACE to absl when we have HAVE_ABSEIL, or define it to absl::otel_v1. By this way we can keep ABI compatibility and use the internal and external abseil at the same time without break the ABI.

Thanks for your analysis @owent. This would be helpful.

owent commented:
@lalitb @marcalff Could you please review this PR again? I have moved base64 codes into sdk and there are a lot of changes, so I removed `od-to-merge` label. I think we can add it back after the new changes are reviewed.

lalitb commented:
> @lalitb @marcalff Could you please review this PR again? I have moved base64 codes into sdk and there are a lot of changes, so I removed `od-to-merge` label. I think we can add it back after the new changes are reviewed.

@owent  - Seems you missed my comment here - https://github.com/open-telemetry/opentelemetry-cpp/pull/2163#discussion_r1284985513 . Actually, I didn't ask for moving base64 from otlp to sdk :)

owent commented:
> > @lalitb @marcalff Could you please review this PR again? I have moved base64 codes into sdk and there are a lot of changes, so I removed `od-to-merge` label. I think we can add it back after the new changes are reviewed.
> 
> @owent - Seems you missed my comment here - [#2163 (comment)](https://github.com/open-telemetry/opentelemetry-cpp/pull/2163#discussion_r1284985513) . Actually, I didn't ask for moving base64 from otlp to sdk in my subsequent comments. But if you want, we can keep it in sdk for now, and have a separate PR to move back to otlp. Rest of the PR looks good.

Thanks. I can move it back to otlp exporter later in another PR.
